engine/src/main/java/org/terasology/engine/world/chunks/localChunkProvider/LocalChunkProvider.java

```diff
@@ -437,10 +406,114 @@ private boolean isChunkReady(Chunk chunk) {
         return chunk != null && chunk.isReady();
     }
 
-    // TODO: move loadingPipeline initialization into constructor.
+    public void notifyRelevanceChanged() {
+        loadingPipeline.notifyUpdate();
+    }
+
+    private void updateList() {
+        chunksInRange.removeIf(x -> !relevanceSystem.isChunkInRegions(x) || isChunkReady(x));
+        relevanceSystem.neededChunks()
+                .filter(pos -> !chunksInRange.contains(pos))
+                .forEach(chunksInRange::add);
+        chunksInRange.sort(relevanceSystem.createChunkPosComparator().reversed());
+    }
+
+    private boolean checkForUpdate() {
+        Collection<ChunkRelevanceRegion> regions = relevanceSystem.getRegions();
+        if (lastRegions == null || regions.size() != lastRegions.length) {
+            lastRegions = regions.stream().map(ChunkRelevanceRegion::getCurrentRegion).toArray(BlockRegion[]::new);
+            return true;
+        }
+        int i = 0;
+        boolean anyChanged = false;
+        for (ChunkRelevanceRegion region : regions) {
+            if (!lastRegions[i].equals(region.getCurrentRegion())) {
+                lastRegions[i].set(region.getCurrentRegion());
+                anyChanged = true;
+            }
+            i++;
+        }
+        return anyChanged;
+    }
+
+    /**
+     * Loads a chunk if possible, otherwise generates it.
+     *
+     * @return The chunk at `pos`, ready for submitting to the ChunkProcessingPipeline.
+     */
+    private Chunk genChunk(Vector3ic pos) {
+        ChunkStore chunkStore = storageManager.loadChunkStore(pos);
+        Chunk chunk;
+        if (chunkStore == null) {
+            EntityBufferImpl buffer = new EntityBufferImpl();
+            chunk = new ChunkImpl(pos, blockManager, extraDataManager);
+            generator.createChunk(chunk, buffer);
+            generateQueuedEntities.put(chunk.getPosition(), buffer.getAll());
+        } else {
+            chunk = chunkStore.getChunk();
+        }
+        return chunk;
+    }
+
+    private Flux<Chunk> chunkFlux() {
+        Set<Vector3ic> currentlyProcessing = new HashSet<>();
+        return Flux.create(sink -> sink.onRequest(numChunks -> {
+            // Figuring out the positions to generate needs to be synchronized
+            List<Vector3ic> positionsPending = new ArrayList<>((int) numChunks);
+            synchronized (this) {
+                if (checkForUpdate()) {
+                    updateList();
+                }
+
+                while (positionsPending.size() < numChunks && !chunksInRange.isEmpty()) {
+                    Vector3ic pos = chunksInRange.remove(chunksInRange.size() - 1);
+                    if (currentlyProcessing.contains(pos) || loadingPipeline.isPositionProcessing(pos)) {
+                        continue;
+                    }
+
+                    positionsPending.add(pos);
+                    currentlyProcessing.add(pos);
```

Comment:
Important thing to know (for newbies like me to the Flux API): This emitter function we're passing to Flux.create (on line 460) is called with a new sink _for each subscriber._

That's how we're managing to add multiple chunk-processing subscribers to this without having each subscriber process the _same_ chunks.

Reply:
I agree, named methods would be good and I think `currentlyProcessing` should just be a member variable. I'll change that, and add some comments explaining when things get called too, since I agree that it's not very intuitive.